tests/pass_by_ref_test.py

```diff
@@ -1,145 +1,358 @@
+                bytes_cursor.set_index(fourteen),
+                Log(bytes_cursor.load()),
+                int_cursor.set_index(fifteen),
+                Log(Itob(int_cursor.load())),
+                bytes_cursor.set_index(sixteen),
+                Log(bytes_cursor.load()),
+                leet,
+            )
+        )
     )
 
 
-def sub_even():
+def empty_scratches():
+    cursor = DynamicScratchVar()
+    i1 = ScratchVar(TealType.uint64, 0)
+    i2 = ScratchVar(TealType.uint64, 2)
+    i3 = ScratchVar(TealType.uint64, 4)
+    s1 = ScratchVar(TealType.bytes, 1)
+    s2 = ScratchVar(TealType.bytes, 3)
+    s3 = ScratchVar(TealType.bytes, 5)
     return Seq(
-        Pop(recursiveIsEven(Int(1000))),
-        recursiveIsEven(Int(1001)),
+        cursor.set_index(i1),
+        cursor.store(Int(0)),
+        cursor.set_index(s1),
+        cursor.store(Bytes("")),
+        cursor.set_index(i2),
+        cursor.store(Int(0)),
+        cursor.set_index(s2),
+        cursor.store(Bytes("")),
+        cursor.set_index(i3),
+        cursor.store(Int(0)),
+        cursor.set_index(s3),
+        cursor.store(Bytes("")),
+        Int(42),
     )
 
 
-if not OLD_CODE_ONLY:
-    #### TESTS FOR NEW PyTEAL THAT USES PASS-BY-REF / DYNAMIC
-    @Subroutine(TealType.none)
-    def logcat_dynamic(first: ScratchVar, an_int):
-        return Seq(
-            first.store(Concat(first.load(), Itob(an_int))),
-            Log(first.load()),
-        )
+@Subroutine(TealType.uint64)
+def oldfac(n):
+    return If(n < Int(2)).Then(Int(1)).Else(n * oldfac(n - Int(1)))
 
-    def sub_logcat_dynamic():
-        first = ScratchVar(TealType.bytes)
-        return Seq(
-            first.store(Bytes("hello")),
-            logcat_dynamic(first, Int(42)),
-            Assert(Bytes("hello42") == first.load()),
-            Int(1),
-        )
 
-    def wilt_the_stilt():
-        player_score = DynamicScratchVar(TealType.uint64)
+CASES = (
+    sub_logcat_dynamic,
+    swapper,
+    wilt_the_stilt,
+    sub_mixed,
+    lots_o_vars,
+    empty_scratches,
+)
 
-        wilt = ScratchVar(TealType.uint64, 129)
-        kobe = ScratchVar(TealType.uint64)
-        dt = ScratchVar(TealType.uint64, 131)
 
-        return Seq(
-            player_score.set_index(wilt),
-            player_score.store(Int(100)),
-            player_score.set_index(kobe),
-            player_score.store(Int(81)),
-            player_score.set_index(dt),
-            player_score.store(Int(73)),
-            Assert(player_score.load() == Int(73)),
-            Assert(player_score.index() == Int(131)),
-            player_score.set_index(wilt),
-            Assert(player_score.load() == Int(100)),
-            Assert(player_score.index() == Int(129)),
-            Int(100),
-        )
+@pytest.mark.skipif(not STABLE_SLOT_GENERATION, reason="cf. #199")
+@pytest.mark.parametrize("pt", CASES)
+def test_teal_output_is_unchanged(pt):
+    assert_new_v_old(pt, 6)
+
+
+def test_pass_by_ref_guardrails():
```

Comment:
Can you consider breaking this test up into multiple functions? There appears to be multiple distinct tests in here, but it's currently hard to distinguish between them since they are all interlaced.

Reply:
👍 - will add cases/tests to other unit test files.